caddy/watch.go

```diff
@@ -0,0 +1,57 @@
+package caddy
+
+import (
+	"github.com/dunglas/frankenphp"
+	"github.com/dunglas/frankenphp/watcher"
+	"path/filepath"
+	"strings"
+)
+
+type watchConfig struct {
+	// Directory that should be watched for changes
+	Dir string `json:"dir,omitempty"`
+	// Whether to watch the directory recursively
+	IsRecursive bool `json:"recursive,omitempty"`
+	// The shell filename pattern to match against
+	Pattern string `json:"pattern,omitempty"`
+}
+
+func applyWatchConfig(opts []frankenphp.Option, watchConfig watchConfig) []frankenphp.Option {
+	return append(opts, frankenphp.WithFileWatcher(
+		watcher.WithWatcherDir(watchConfig.Dir),
+		watcher.WithWatcherRecursion(watchConfig.IsRecursive),
+		watcher.WithWatcherPattern(watchConfig.Pattern),
+	))
+}
+
+func parseWatchConfigs(filePatterns []string) []watchConfig {
+	watchConfigs := []watchConfig{}
+	for _, filePattern := range filePatterns {
+		watchConfigs = append(watchConfigs, parseWatchConfig(filePattern))
+	}
+	return watchConfigs
+}
+
+// TODO: better path validation?
+// for the one line short-form in the caddy config, aka: 'watch /path/*pattern'
+func parseWatchConfig(filePattern string) watchConfig {
+	watchConfig := watchConfig{IsRecursive: true}
+	dirName := filePattern
+	splitDirName, baseName := filepath.Split(filePattern)
+	if filePattern != "." && filePattern != ".." && strings.ContainsAny(baseName, "*.[?\\") {
+		dirName = splitDirName
+		watchConfig.Pattern = baseName
+		watchConfig.IsRecursive = false
+	}
+	if strings.Contains(filePattern, "**/") {
+		dirName = strings.Split(filePattern, "**/")[0]
+		watchConfig.Pattern = strings.Split(filePattern, "**/")[1]
+		watchConfig.IsRecursive = true
+	}
+	watchConfig.Dir = dirName
+	if dirName != "/" {
+		watchConfig.Dir = strings.TrimRight(dirName, "/")
+	}
+
+	return watchConfig
+}
```

Comment:
This whole function feels a bit weird. Maybe add some comments, sanitization, etc?

```suggestion
func parseWatchConfig(filePattern string) watchConfig {
    var config watchConfig
    config.IsRecursive = true

    // Handle empty filePattern
    if filePattern == "" {
        config.Dir = "."
        return config
    }

    // Clean the filePattern
    filePattern = filepath.Clean(filePattern)

    // Check if filePattern contains "**/"
    if strings.Contains(filePattern, "**/") {
        parts := strings.SplitN(filePattern, "**/", 2)
        config.Dir = parts[0]
        config.Pattern = parts[1]
        config.IsRecursive = true
    } else {
        dir := filepath.Dir(filePattern)
        base := filepath.Base(filePattern)

        // Check if base contains glob characters
        if strings.ContainsAny(base, "*?[") {
            config.Dir = dir
            config.Pattern = base
            config.IsRecursive = false
        } else {
            config.Dir = filePattern
            config.IsRecursive = true
        }
    }

    // Trim trailing slashes from Dir unless it's root
    if config.Dir != "/" {
        config.Dir = strings.TrimRight(config.Dir, "/")
    }

    return config
}
```

untested example ^

Even then, it feel a bit weird. Like imagine the following directory layout:

```
my-app/
├─ public/
│  ├─ favicon.ico
│  ├─ index.html
│  ├─ robots.txt
├─ src/
│  ├─ index.css
│  ├─ index.php
├─ vendor/
│  ├─ liba/
│  │  ├─ src/
│  │  |  ├─ public/
│  ├─ libb/
│  │  ├─ src/
├─ .gitignore
├─ README.md
```

and I pass the following pattern: `./vendor/**/src/**/public/*.php`

This implementation seems like it would capture everything in the vendor directory, which is not what is specified.

Reply:
I'm not very comfortable with this function either. TBH, I would prefer a very basic option just exposing what is supported by the underlying library for now (which may be a way to exclude directories).

Reply:
This will make tests fail, but I'll refactor it to make the intentions more clear 👍 . 
`./vendor/**/src/**/public/*.php` will not match anything, `**/` is supposed to just be used once. I'll throw an error if it's there twice.

The idea was to keep it simple so we could easily switch to a different watcher implementation (that's why I also removed regex). Since `filepath.Match` doesn't support the doublestar glob '**' though, one `**/` followed by a filepattern without '/' is allowed. The main use case is probably just something like `/app/**/*.php`. 

The [filepath.Abs](https://pkg.go.dev/path/filepath#Abs) in `watch_options.go` will handle `""` case same as `"."`  and call `Clean()` on the result, so it's not necessary to validate it here as well.

Reply:
> **/ is supposed to just be used once.

I have a project right now that uses it more than once in a package.json:

`public/**/css/**/src/**/*.scss`

It needs to match `public/web/css/home/header/src/component/buttons/button.scss`, for example.

For my custom restarter on another project, I have it watching the following directories (translated to this globbing format):

- src/**/*.php
- vendor/my-org/**/src/**/*.php

where the packages are symlinked modules from other directories.

I suspect this is rather common on larger projects, especially ones like ours at work where we have custom symfony bundles living in other parts of the codebase.

Reply:
I'll look into supporting a multi `**` pattern if that is something people need, will definitely need some testing. I don't think that `edant/watcher` supports following symlinks though. At least it's not an option that can be passed.

Reply:
Symlinks should always followed on Linux. I don't know what the behavior is in Windows and MacOS (Windows doesn't really have a proper symbolic link concept, and macOS doesn't document the behavior around symlinks for fsevents AFAICT).

But we don't currently have a proper test for following symlinks and out own documentation on that is missing, so we can fix that.

Reply:
Well, to be sure, event paths are always reported by their real path (symlinks followed), watched paths are always followed to their real path before being watched, and that is always true, on every platform.

But I don't know exactly what happens in some of the edge cases where, say, a new directory spawns within a watched directory, but the new symlink-to-a-directory is outside of the watched paths.

Reply:
Thank you @e-dant . I'll do some trials with symlinks then.

Multi-globbing (`public/**/css/**/src/**/*.scss`) is now supported @withinboredom

Reply:
@dunglas the underlying library doesn't support any pattern matching, we would always have to watch all files in a directory. I refactored this to make the intentions clearer and added a lot of tests. I think the only way to be flexible with the underlying library is if we do the pattern matching ourselves.

Reply:
@withinboredom can we do symlink support in a future PR? It should probably be optional and would require resolving all symlinks to their absolute paths (aka walking through all files)